docs/design/2021-03-09-security-enhanced-mode.md

```diff
@@ -0,0 +1,216 @@
+    * [Functional Tests](#functional-tests)
+    * [Scenario Tests](#scenario-tests)
+    * [Compatibility Tests](#compatibility-tests)
+    * [Benchmark Tests](#benchmark-tests)
+* [Impacts & Risks](#impacts--risks)
+* [Investigation & Alternatives](#investigation--alternatives)
+* [Unresolved Questions](#unresolved-questions)
+
+## Introduction
+
+This document was created to discuss the design of Security Enhanced Mode. It comes from the DBaaS requirement that `SUPER` users must not be able to perform certain actions that could comprimise the system.
+
+### Terminology
+
+* **Configuration Option**: The name of a variable as set in a configuration file.
+* **System Variable** (aka sysvar): The name of a variable that is set in a running TiDB server using the MySQL protocol.
+* **Super**: The primary MySQL "admin" privilege, which is intended to be superseded by MySQL’s "dynamic" (fine-grained) privileges starting from MySQL 8.0.
+
+## Motivation or Background
+
+Currently the MySQL `SUPER` privilege encapsulates a very large set of system capabilities. It does not follow the best practices of allocating _fine grained access_ to users based only on their system-access requirements.
+
+This is particularly problematic in a DBaaS scenario such as TiDB Cloud where the `SUPER` privilege has elements that are required by both end users (TiDB Cloud Customers) and system operations (PingCAP SREs).
+
+The design of Security Enhanced Mode (SEM) takes the approach of:
+
+1. Restricting `SUPER` to a set of capabilities that are safe for end users.
+2. Implementation of dynamic privileges ([issue #22439](https://github.com/pingcap/tidb/issues/22439)).
+
+This approach was requested by product management based on the broad "in the wild" association of `SUPER` as "the MySQL admin privilege". Thus, proposals to create a new lesser-`SUPER` privilege have already been discussed and rejected.
+
+The design and name of "Security Enhanced" is inspired by prior art with SELinux and AppArmor.
+
+## Detailed Design
+
+A boolean option called `EnableEnhancedSecurity` (default `FALSE`) will be added as a TiDB configuration option.  The following subheadings describe the behavior when `EnableEnhancedSecurity` is set to `TRUE`.
+
+### System Variables
+
+The following system variables will be hidden:
+
+* variable.TiDBDDLSlowOprThreshold,
+* variable.TiDBAllowRemoveAutoInc,
+* variable.TiDBCheckMb4ValueInUTF8,
+* variable.TiDBConfig,
+* variable.TiDBEnableSlowLog,
+* variable.TiDBEnableTelemetry,
+* variable.TiDBExpensiveQueryTimeThreshold,
+* variable.TiDBForcePriority,
+* variable.TiDBGeneralLog,
+* variable.TiDBMetricSchemaRangeDuration,
+* variable.TiDBMetricSchemaStep,
+* variable.TiDBOptWriteRowID,
+* variable.TiDBPProfSQLCPU,
+* variable.TiDBRecordPlanInSlowLog,
+* variable.TiDBRowFormatVersion,
+* variable.TiDBSlowQueryFile,
+* variable.TiDBSlowLogThreshold,
+* variable.TiDBEnableCollectExecutionInfo,
+* variable.TiDBMemoryUsageAlarmRatio,
```

Comment:
I wish to talk about `wait_timeout` and `interactive_timeout` on the cloud.


1. The TiDB's default `wait_timeout` is unlimited, but MySQL and Aurora MySQL's default are 8 hours.


2. Aurora does not follow a too low  `wait_timeout` and `interactive_timeout`, to avoid users' misconfiguration.

> Aurora evaluates the minimum value of interactive_timeout and wait_timeout, then uses that minimum as the timeout to end all idle sessions, both interactive and noninteractive.

Reply:
.. but in Aurora's case I'm sure they let users modify these values? They just don't permit very low values (but use the semantic of ignore instead of error, which I assume is for compatibility reasons).

I think we can lower the defaults, but I don't think it needs to be grouped in with security.

Reply:
I think ignoring in this context is because some connector or connection pool might set a low value, and Aurora still wants to work with it out of the box.

Unless there is a known compatibility reason (or likely reason), ignore is a risky semantic because users may think they've made a change, but it does not take effect. For example see [my comment on temporary tables](https://github.com/pingcap/tidb/pull/22860#issuecomment-785231707) and [read-only transactions](https://github.com/pingcap/tidb/issues/22658).

So ignoring should be the **exception** and **not the rule**. In TiDB it has historically been the rule, which is problematic.